pylint-dev/pylint
# Add a new ``declare-non-slot`` error code

## Type of Changes



|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description



Refs #9499 

Implement `declare-non-slot` which reports an error when a class has a `__slots__` member and a type hint on the class is not present in `__slots__`.

## Details

**Tests:**

- Add test case which should report the error code.
- Add test case where the error is suppressed due to `__dict__` in `__slots__`.

- Add test to detect annotation not in any base class's `__slots__`.

- Add test against false-positive when base doesn't have `__slots__` (since there is still then an instance `__dict__` thanks to base)

- Add test against false-positive when base has `__slots__` with `__dict__` entry.

- Ignore `declare-non-slots` for `regression_5479.py` which is for `assigning-non-slot`.

**Implementation:**

Implement check for `declare-non-slot` in `class_checker.py` and 2 new helper methods:

- Add method `_check_declare_non_slot` to report `declare-non-slot` error. This method checks `node` and all bases for a valid `__slots__`, gathering the names in all `__slots__` found. If `node` has an annotation not in any `__slots__`, then `declare-non-slot` is reported.

- Add helper method `_has_valid_slots` which returns `True` if a valid `__slots__` is found on a `ClassDef` (re-use logic from `_check_slots`).

- Refactor `_check_redefined_slots` to split out logic for getting `__slots__` names into `_get_classdef_slots_names` helper, since the same logic can be used by both functions.

**Some thoughts:**

- I thought this check belonged in the class checker instead of the dataclass checker since the problematic code is problematic even without the dataclass decorator. There may be other cases where the class annotations and __slots__ items should match, and making this check dataclass-specific would miss those cases.
- I went with `declare-non-slot` over `define-non-slot` as annotating a class feels more like a declaration that such a member exists, rather than a definition of that member.
- I made some changes to existing code to factor out some logic into a helper function `_get_classdef_slots_names` as this seemed the cleanest way to re-use existing logic. I hope this is ok.
- I've checked that the existing tests still pass and have run the primer tests with `python3 -m pytest -m primer_stdlib --primer-stdlib`.


**To Do:**

- Implement `missing-slot-annotation` from #9499
- ~Add documentation (good/bad code examples)~
- ~Add changelog/towncrier entry~
